### DISCUSSION TITLE
WinUI 3 specific build

### DISCUSSION THREAD

**Discussion by brabebhin** [ffmpeginteropx/FFmpegInteropX#343] [General]
I have been thinking about this for a while, how would we approach a winUI specific build?
It seems the code changes are quite extensive in some area, even requiering different nuget packages installed, and different idl file (this last one could theoretically be delt with using compiler directives)

Should we try to use a shared project?
Try to diverge the code base? 
I think in the long run, winUI allows for more powerful builds (I think we could technically link libraries which aren't available in UWP) and features.
Perhaps we should consider simply forking the code and backport UWP supported features?
Or try to support UWP for the forseeable future? For how long? We are 2 people after all, so we can't really keep 2 code bases in sync forever.

**Comment by lukasf** [comment]
I think it should be possible to get both builds from a single code base using two different build configurations. IIRC in the unit test branch, I already introduced a WinUI build (need to look that up, it's been a while). By using conditions in the .vcxproj for the NuGet references, we can switch to only include the WinUI stuff in the corresponding build configuration. This could be a bit messy when updating the NuGet packages (because that will probably overwrite our confitions). But still if that works, it would be the cleanest solution.

For me the bigger question is about the output NuGet packages. Should we ship one package that contains both the WinUI and UWP versions? Or should we ship two separate packages? Should we create a WinUI/Win32 FFmpeg build, to be used by FFmpegInteropX(WinUI)? I guess a separate FFmpeg build would be required if we want to allow file names as source.

Once we have the dual build setup done, I do not see much additional work maintaining both UWP and WinUI. After all, most of it is identical, since the core media classes are the same. Only dispatcher handling and UI/Display environment stuff is different, and that is not something we use a lot, luckily.

**Comment by brabebhin** [comment]
All very good points. I guess we can sleep on it a few more months until winUI stabilizes and matures. I'm a little concerned about the cs/winrt project though, it doesn't seem very lively.

**Reply by lukasf** [reply]
True, I also do not see us in a hurry on this.

By the way, I have started a new project using Uno/WinUI as an experiment. When I tested Uno last time, the performance was totally unacceptable. But things have improved a whole lot over there. It took me quite some optimizations for the Android version, but now I am pretty impressed by the performance. Not like native performance, but a lot better than what I expected. Uno is now definitely an option for me, if I want to go cross-platform with an app.

The WinUI experience is so-so. A lot of things work really well. The new controls are nice. Performance is okay (it would be better with native AOT ofc). But then there are some mind blowing bugs (like 100% reproducible access violation crashes on drag+drop on grouped list views) which do not even seem to be worked on. WTF?! On the other hand I read that Windows Explorer will be partly moved to WinUI. And in the experimental version, I see new, advanced controls like annotated seek bar. Huh what?? It is really hard to understand what is going on there. Their strategy is totally messed up. Things not fit together at all.

Anyways, yeah no need for us to hurry. They need to get their sh*t together and that will still take some time.

**Reply by brabebhin** [reply]
I also started porting my old uwp media player to winUI and i will open source it in the coming months, once I'm happy with the state of the codebase. Some classes are over 12 years old, when i was in my 2nd year of university lol. Terrible code in there. 

The biggest performance gain is file access. Seriously. Windows.Storage namespace is so ridiculously slow compared to net core apis it isn't even funny. That alone almost makes up for lack of aot. 

On the other hand, lots of weird bugs indeed. The windowing api is all over the place (like can't set minimum window size lol or requiring nasty interop for basic stuff). Full screen seems broken. Advanced controls look cool. Until you mix their more advanced properties and get some very funny results. Some event handlers throw stack overflow exceptions if you attach handlers on the wrong thread. Lots of localization bugs. 

I'm scared at the prospect of a winui file explorer. The taskbar and start menu are total jokes in w11, hopefully the new explorer doesn't cut too many of its features.

In surprised uno is in a usable state. I've been looking for a productive alternative to kotlin / compose for android, but maui and net core don't seem to tackle fundamental issues (particularly native bindings). Maui is also a very bad joke when it comes to performance and stability. Somehow worse than winui.

**Reply by lukasf** [reply]
Yeah, my new project is actually a rewrite of a music player app I done 15 years ago. And I was amazed as well about the performance of file APIs. The initial scan of the music library was REALLY slow back then. And even just searching for changes took nearly a minute, without even reading a single file. Of course, my current system is dramatically faster, and it has SSD. But the speed of library parsing is really mind blowing, and a lot of that comes from the faster APIs. If I disable virus guard for the music folder, I can scan 300gb of music in 7 seconds!! Including full tag parsing, cover reading and populating a library database. Scanning the 300gb folder for changes takes 0.5 sconds. That's really sick ^^ So fast that it does not even make sense to show update progress.

The WinUI windowing APIs are indeed out of place. I had a lot of trouble with getting transparent background (mica/acrylic) and custom app bar working. There are two places where you can do these things, and don't you even dare mixing the two of them, you will end up with a complete mess. But then, some features are only available in one of the APIs, and some only in the other. D'oh

The WinUI 3 TreeView is so full of bugs, it's ridiculous.

I did not mean to say that uno is great for building android UIs. Nothing beats a native android UI. So if you want to build a great app for android, with a great and fast UI, you have to go with compose or other native UI stacks. But if your goal is to go cross platform easily without much additional work, and you accept compromises on UI (performance is inferior, animations are limited), then uno could be an option.

**Comment by brabebhin** [comment]
I'm not that impressed with the native android tools either, the xml based ui is akin to spaghetti, even the most basic stuff requires a lot of glue code and magic string matching.

The compose stuff looks cool but it is simply not mature enough and seems to have started on the wrong foot already.

Witch brings me to xaml. Xaml is very efficient for ui, but it comes with the difficulty of consuming native libraries. I feel that flutter has a better approach to the issue, by allowing you to mix code from different languages in the same project. I'd honestly want visual studio to support compiling kotlin and java and automatically generate the bindings for you. Sort of like the one winRT times...

One can only dream of such tooling support.

**Reply by lukasf** [reply]
I was also pretty shocked by the android APIs. What a horrible mess. They are deprecating and replacing APIs like maniacs. Then trying to fix the mess up with this compat layer. Oh man. WinRT APIs are miles ahead in nearly every aspect.

I wish that MS would finally open source WinUI, like they promised long ago. Then maybe someday someone (probably not microsoft) might be able cross-compile that as native lib (Android/iOS) and add a renderer in the system's 3D tech. The main reason why performance of Uno and similar techs is slow is because they hook into the system's GUI framework, emulating native controls and taking part in their layout system. This causes lots and lots of interop overhead. If they'd instead run as a pure 3D app with their own, fully native code layout+rendering system, performance would be way beyond what we have now. That's what I am dreaming of... ^^
